File: cpp/leetcode/矩阵置零/1.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();

        bool row0 = false;
        bool col0 = false;

        for (int i = 0; i < m; ++i) {
            if (matrix[i][0] == 0) {
                col0 = true;
                break;
            }
        }
        for (int j = 0; j < n; ++j) {
            if (matrix[0][j] == 0) {
                row0 = true;
                break;
            }
        }

        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        if (row0) {
            for (int j = 0; j < n; ++j) {
                matrix[0][j] = 0;
            }
        }
        if (col0) {
            for (int i = 0; i < m; ++i) {
                matrix[i][0] = 0;
            }
        }
    }
};

int main() {
    Solution solution;
    vector<vector<int>> matrix = {{0, 1, 2, 0}, {3, 4, 5, 2}, {1, 3, 1, 5}};
    solution.setZeroes(matrix);
    cout << "[";
    for (const vector<int>& nums : matrix) {
        cout << "[";
        for (int i = 0; i < nums.size(); ++i) {
            cout << nums[i];
            if (i < nums.size() - 1) {
                cout << ", ";
            }
        }
        cout << "]";
        if (nums != matrix.back()) {
            cout << ", ";
        }
    }
    cout << "]" << endl;
}

File: cpp/leetcode/矩阵置零/1.py
from typing import List

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> List[List[int]]:
        m, n = len(matrix), len(matrix[0])
        row0, col0 = False, False

        for i in range(m):
            if matrix[i][0] == 0:
                col0 = True
                break
        for j in range(n):
            if matrix[0][j] == 0:
                row0 = True
                break

        for i in range(1,m):
            for j in range(1,n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        for i in range(1,m):
            for j in range(1,n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        if row0:
            for j in range(n):
                matrix[0][j] = 0

        if col0:
            for i in range(m):
                matrix[i][0] = 0

        return matrix

if __name__ == "__main__":
    solution = Solution()
    matrix = [[0,1,2,0],[3,4,5,2],[1,3,1,5]]
    result = solution.setZeroes(matrix)
    print(result)